### ISSUE TITLE
Naval Warefare: Soldiers on ships are sometimes messed up

### ISSUE THREAD

**Issue by frankystone** [widelands/widelands#5973]
Somehow i got into a situation where one war ship can't be filled up with soldiers anymore. Additionally there are no destinations possible for this ship. After reloading the save game it works as expected, so i can't  provide a save game. Unfortunately i have also no replay.

In this situation of the "lost" war ship i noticed also some messed up soldiers counts on war ships:
There where two war ships, the "lost" one and another filled with 6 Soldiers. While i was trying to navigate the "lost" ship to a port i anchored the second war ship few fields away from that port. While playing around with the amount of soldiers in the "lost" war ship the anchored ship showed suddenly 7 soldiers on bord.

Will try to reproduce, but this can take a while.

**Comment by frankystone** [comment]
An example image, the capacity number says 6 Soldiers, the amount of soldiers in the list on top says 7 soldiers: 
![shot0006](https://github.com/widelands/widelands/assets/6730556/80a513b2-8c63-412b-a14f-23c51fda8070)

**Comment by frankystone** [comment]
Don't know if this is related to the messed up soldier list:

If you try to attack the port the game crashes with:
~~~~
widelands: /home/kaputtnik/Quellcode/widelands-repo/widelands/src/logic/map.h:715: static Widelands::MapIndex Widelands::Map::get_index(const Widelands::Coords &, const int16_t): Assertion `0 <= c.y' failed.
Abgebrochen (Speicherabzug geschrieben)
~~~~

But if you navigate the ship back to Luvavum and make sure there are only 6 soldiers on board, attacking that port will result in the same crash.

[assertion.zip](https://github.com/widelands/widelands/files/11926695/assertion.zip)

**Comment by tothxa** [comment]
> An example image, the capacity number says 6 Soldiers, the amount of soldiers in the list on top says 7 soldiers: 
> 
> And this ship just had found a foreign port, so the amount of 7 soldiers in the list can't be made by misclicks or so.

Which commit are you using? This sounds like #5948 which should have been fixed by #5953 in https://github.com/widelands/widelands/commit/716a0ec0a5d20b70feb97c0ba78be1dfde108b6c

**Comment by tothxa** [comment]
> Don't know if this is related to the messed up soldier list:
> 
> If you try to attack the port the game crashes with:

It's unrelated. I was able to debug it. (RelWithDebInfo with -O2 → seemingly nonsense line in backtrace → made me curious :)))

The portspace has 0 y coordinate, so `Game::random_location()` used for scattering soldiers gives some negative coordinates. I guess it should normalise the coordinates before returning them to be safe. (`findpath()` normalises its endpoints, so most users were safe already.)

**Comment by frankystone** [comment]
> > An example image, the capacity number says 6 Soldiers, the amount of soldiers in the list on top says 7 soldiers:
> > And this ship just had found a foreign port, so the amount of 7 soldiers in the list can't be made by misclicks or so.
> 
> Which commit are you using? This sounds like #5948 which should have been fixed by #5953 in [716a0ec](https://github.com/widelands/widelands/commit/716a0ec0a5d20b70feb97c0ba78be1dfde108b6c)
> 
> _edit: I see the save is with a newer commit. Is it possible that you started the game and filled this ship with an earlier commit?_

No. I checked out the Tournament 2023 tag, and just started playing.

**Comment by frankystone** [comment]
There is another bug, probably related to #5948 :
1. Refit a ship to a warship
2. the soldiers list on top shows soldiers inside the ship and awaited soldiers
3. reduce the amount of soldiers one by one by hitting the down arrow
4. if the list shows only one awaiting soldier clicking the down arrow doesn't have an effect
5. an additional click reduces the amount by two then

**Comment by frankystone** [comment]
Closing as fixed by #6006
